File: tests/lwrtest.cpp
#include <stdlib.h>
#include <assert.h>
#include <math.h>
#include <iostream>
#include <fstream>
#include <vector>
#include <utility>
#include <iterator>
#include <sstream>
#include "lwr.h"
#include "common.h"

using namespace std;

void parsefloats(const string &s, vector<float> &v) {
	stringstream ss(s);
	float f;
	int c = 1;
	while (!ss.eof()) {
		if (!(ss >> f)) {
			if (ss.eof()) {
				break;
			}
			cerr << "error parsing field " << c << " in " << ss.str() << endl;
			exit(1);
		}

		v.push_back(f);
		c++;
	}
}

void parseline(const string &line, floatvec &x, floatvec &y, float &dt) {
	vector<string> parts;
	vector<float> x1, y1;
	
	split(line, ";", parts);
	assert(parts.size() == 3);
	parsefloats(parts[0], x1);
	parsefloats(parts[1], y1);
	stringstream ss(parts[2]);
	if (!(ss >> dt)) {
		cerr << "error parsing " << parts[2] << endl;
		exit(1);
	}
	x = x1;
	y = y1;
}

int main(int argc, char *argv[]) {
	string line;
	int nnbrs;
	lwr *m = NULL;
	vector<double> diffs;
	
	if (argc != 5) {
		cerr << "usage: lwrtest <method> <num neighbors> <training file> <test file>" << endl;
		exit(1);
	}
	
	stringstream ss(argv[2]);
	if (!(ss >> nnbrs)) {
		cerr << "nnbrs parse error" << endl;
		exit(1);
	}
	
	ifstream trainf(argv[3]);
	if (!trainf) {
		cerr << "couldn't open " << argv[3] << endl;
		exit(1);
	}
	
	while (getline(trainf, line)) {
		floatvec x, y;
		float dt;
		parseline(line, x, y, dt);
		if (!m) {
			m = new lwr(x.size(), y.size(), nnbrs);
		}
		for(int i = 0; i < y.size(); ++i) {
			y[i] = (y[i] - x[i]) / dt;
		}
		m->add(x, y);
	}
	
	trainf.close();
	
	ifstream testf(argv[4]);
	if (!testf) {
		cerr << "couldn't open " << argv[4] << endl;
		exit(1);
	}
	
	while (getline(testf, line)) {
		floatvec x, y, dy, py;
		float dt;
		parseline(line, x, y, dt);
		dy.resize(y.size());
		py.resize(y.size());
		m->predict(x, dy, argv[1][0], false);
		for(int i = 0; i < py.size(); ++i) {
			py[i] = x[i] + dy[i] * dt;
		}
		cout << "predict: " << py << endl;
		cout << "actual:  " << y << endl;
		diffs.push_back(sqrt(py.distsq(y)));
	}
	
	if (diffs.size() == 0) {
		cout << "No predictions" << endl;
		return 0;
	}
	
	double mean = 0.0, std = 0.0;
	for (int i = 0; i < diffs.size(); ++i) {
		mean += diffs[i];
	}
	mean /= diffs.size();
	
	for (int i = 0; i < diffs.size(); ++i) {
		std += pow(diffs[i] - mean, 2);
	}
	std = sqrt(std / diffs.size());
	
	cout << "MEAN: " << mean << endl;
	cout << "STD:  " << std << endl;
	return 0;
}

File: tests/lwrtest.py
from __future__ import print_function
import sys, os, math, random
from subprocess import *
sys.path.append('.')
from lwr import LWR
import numpy as np

def dist(x1, x2):
	d = x1 - x2
	return math.sqrt(np.dot(d, d))

def toarray(line):
	return np.array([float(v) for v in line.split()])
	
def regress(nnbrs, text):
	p = Popen(['./lwrtest', str(nnbrs)], stdin=PIPE, stdout=PIPE)
	out = p.communicate(text)[0]
	return toarray(out)

if len(sys.argv) != 3:
	print('usage: lwrtest.py <num neighbors> <data file>', file=sys.stderr)
	sys.exit(1)

m = None
nnbrs = int(sys.argv[1])
diffs = []
nulldiffs = []
training = True
for line in open(sys.argv[2]):
	if line.startswith('$'):
		training = False
		continue
		
	x, y = map(toarray, line.split(';'))
	if training:
		if not m:
			m = LWR(len(x), len(y), nnbrs)
		m.update(x, y)
	else:
		p = m.predict(x)
		diffs.append(dist(p, y))
		nulldiffs.append(dist(x[:len(y)], y))
	
diffa = np.array(diffs)
print('Mean: ', np.mean(diffa))
print('STD:  ', np.std(diffa))

ndiffa = np.array(nulldiffs)
print('NULL Mean: ', np.mean(ndiffa))
print('NULL STD:  ', np.std(ndiffa))
